### PR TITLE
Fetch health info action

### PR DESCRIPTION
**Pull request by masseyke**
This PR adds FetchHealthInfoCacheAction, which fetches the data sent to the health node by UpdateHealthInfoCacheAction.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/health/node/HealthInfoCache.java [hunk 1]
```diff
@@ -55,8 +55,16 @@ public void clusterChanged(ClusterChangedEvent event) {
         }
     }
 
-    // A shallow copy is enough because the inner data is immutable.
     public Map<String, DiskHealthInfo> getDiskHealthInfo() {
```
Comment (gmarouli) [inline server/src/main/java/org/elasticsearch/health/node/HealthInfoCache.java:58]: I think we do not need this method, `HealthInfo:getHealthInfo` is providing the same information and a client of this method can extract further what they need. What do you think?

File: server/src/main/java/org/elasticsearch/health/node/HealthInfo.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import org.elasticsearch.common.io.stream.Writeable;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * This class wraps all the data returned by the health node, to be consumed by the HealthIndicatorServices.
+ * @param diskInfoByNode A Map of node id to DiskHealthInfo for that node
+ */
+public record HealthInfo(Map<String, DiskHealthInfo> diskInfoByNode) implements Writeable {
```
Comment (gmarouli) [inline server/src/main/java/org/elasticsearch/health/node/HealthInfo.java:22]: I like this structure, it serves the data grouped by indicator which is probably how they will be consumed too :)

File: server/src/test/java/org/elasticsearch/health/node/HealthInfoTests.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+package org.elasticsearch.health.node;
+
+import org.elasticsearch.health.HealthStatus;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.test.EqualsHashCodeTestUtils;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class HealthInfoTests extends ESTestCase {
```
Comment (gmarouli) [inline server/src/test/java/org/elasticsearch/health/node/HealthInfoTests.java:18]: Do you know about this `AbstractWireSerializingTestCase`? You can use it to simplify serialization tests for Writables.

File: server/src/main/java/org/elasticsearch/health/node/HealthInfoCache.java (hunk 1 above)
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/health/node/HealthInfoCache.java:58]: Yeah I initially deleted it and then put it back for convenience. Happy to delete it.

File: server/src/test/java/org/elasticsearch/health/node/HealthInfoTests.java (hunk 3 above)
Comment (masseyke) [reply server/src/test/java/org/elasticsearch/health/node/HealthInfoTests.java:18]: I did not know about that. Checking it out now. Thanks!

Comment (gmarouli) [reply server/src/test/java/org/elasticsearch/health/node/HealthInfoTests.java:18]: Let me know if it works out, I think I have used something like this in the past.

**Comment by gmarouli** [comment]
Thanks for joining me in this disk health efforts! Nice work!

File: server/src/test/java/org/elasticsearch/health/node/FetchHealthInfoCacheActionTests.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static org.elasticsearch.test.ClusterServiceUtils.createClusterService;
+import static org.elasticsearch.test.ClusterServiceUtils.setState;
+import static org.hamcrest.Matchers.equalTo;
+
+public class FetchHealthInfoCacheActionTests extends ESTestCase {
```
Comment (gmarouli) [inline server/src/test/java/org/elasticsearch/health/node/FetchHealthInfoCacheActionTests.java:42]: Nice tests, I will create one of these for the update action too.

File: server/src/test/java/org/elasticsearch/health/node/HealthInfoTests.java (hunk 3 above)
Comment (masseyke) [reply server/src/test/java/org/elasticsearch/health/node/HealthInfoTests.java]: Yeah that worked well -- it required me to provide basically the same code, but tests a lot more. I've pushed it.

Comment (gmarouli) [reply server/src/test/java/org/elasticsearch/health/node/HealthInfoTests.java]: Nice!

File: server/src/test/java/org/elasticsearch/health/node/HealthInfoTests.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            diskInfoByNode.put(randomAlphaOfLengthBetween(10, 100), diskHealthInfo);
+        }
+        return new HealthInfo(diskInfoByNode);
+    }
+
+    @Override
+    public HealthInfo mutateInstance(HealthInfo originalHealthInfo) {
+        Map<String, DiskHealthInfo> diskHealthInfoMap = originalHealthInfo.diskInfoByNode();
+        Map<String, DiskHealthInfo> diskHealthInfoMapCopy = new HashMap<>(diskHealthInfoMap);
+        switch (randomIntBetween(1, 3)) {
```
Comment (gmarouli) [inline server/src/test/java/org/elasticsearch/health/node/HealthInfoTests.java:41]: I was thinking we could simplify this code a bit. We could first do a check if the map is empty and always add an entry, else we can execute the insert, update, remove cases. What do you think?

Comment (masseyke) [reply server/src/test/java/org/elasticsearch/health/node/HealthInfoTests.java:41]: Yeah I can do that

**Pull request closed without merging**